src/cargo/core/workspace.rs

```diff
@@ -459,6 +469,30 @@ impl<'cfg> Workspace<'cfg> {
         Ok(None)
     }
 
+    /// After the root of a workspace has been located, loads the `workspace.metadata` table from
+    /// the workspace root file.
+    ///
+    /// If no workspace was defined, then no changes occur.
+    fn load_workspace_metadata(&mut self) -> CargoResult<()> {
```

Comment:
I would maybe recommend looking into changing this so that it is `fn workspace_config()` that returns the workspace config. This is a duplicate of the code in `find_members`, so it seems reasonable to share that.  You can then remove the `custom_metadata` from the `Workspace`, and only fetch it when building `ExportInfo`.

Reply:
This approach seems to require that the fn workspace_config() take a &mut self, as self.packages.load() requires a &mut self.

I could convert the internal Packages::packages cache into a RefCell or Mutex, but that would seem to make things a bit more complex than necessary.

At least for now, I'll still extract the load_workspace_config() function,, but keep the custom_metadata field as it is.

Reply:
Ah, I see.  The call to `clone` is a bit unfortunate, but it seems like it would be a bit awkward to avoid.